### ISSUE TITLE
Linkage can fail spuriously for newly installed formulae

### ISSUE THREAD

**Issue by mikhailkoliada** [Homebrew/brew#13836]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
: 3.6.0-24-g5524b30
ORIGIN: https://github.com/Homebrew/brew
HEAD: 5524b302d83a54660c49e850b3bc1558371c2fbc
Last commit: 27 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 6f7e54470b33fe838ac4fbccd47d743a8baee15d
Core tap last commit: 24 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_CLEANUP_PERIODIC_FULL_DAYS: 3650
HOMEBREW_MAKE_JOBS: 1
HOMEBREW_NO_AUTO_UPDATE: set
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: single-core 64-bit haswell
Clang: N/A
Git: 2.25.1 => /bin/git
Curl: 7.68.0 => /bin/curl
Kernel: Linux 5.15.0-1019-azure x86_64 GNU/Linux
OS: Ubuntu 20.04.5 LTS (focal)
Host glibc: 2.31
/usr/bin/gcc: 9.4.0
/usr/bin/ruby: N/A
glibc: 2.35
gcc@11: 11.3.0
xorg: N/A

```brew doctor
Your system is ready to brew```
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

actions/runner-images projects [installs](https://github.com/actions/runner-images/blob/main/images/linux/toolsets/toolset-2004.json#L219) zstd from brew on Ubuntu 18.04 and 20.04 (just to make sure it is always present in the system as customers demand it a lot). With homebrew 3.6.0 gcc@11 has become a mandatory dependency for any package, so we decided to add gcc@11 into our list (put it first), and expected zstd's explicit installation to be quietly ignored as it is pulled by gcc itself in anyway (but we still want to keep it listed explicitly, just in case).

### What happened (include all command output)?

After test CI runs we have noticed that brew experienced linkge problems for gcc an it tries to rebuild itself from sources, while it should not. See build [runs](https://github.visualstudio.com/runner-images/_build/results?buildId=128717&view=logs&j=011e1ec8-6569-5e69-4f06-baf193d1351e&t=5431112d-2b61-5a5f-7042-ef698f761043)

### What did you expect to happen?

zstd installation ignored and no linkage problems for gcc@11

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
`brew install gcc@11`
`brew install zstd`
```

**Comment by carlocab** [comment]
I can reproduce it consistently with [this workflow](https://github.com/carlocab/workflow-test/blob/91eec45563f637568733b011fdcdd9b958e9a8b9/.github/workflows/test.yml).

After a fresh `brew install gcc@11`, we get the following failure for `brew linkage --test gcc@11`:
```
Missing libraries:
  unexpected (libgcc_s.so.1)
  unexpected (libquadmath.so.0)
  unexpected (libstdc++.so.6)
  unexpected (libzstd.so.1)
```
This seems to be because after reinstalling, [we see](https://github.com/carlocab/workflow-test/runs/8275464330?check_suite_focus=true#step:6:84):
```
==> Reinstalling 3 dependents with broken linkage from source:
binutils, zstd, gcc@11
```
In my case, `brew` doesn't actually go through reinstalling any of the above formulae from source.

If we run `ls -la "$(brew --cellar gcc@11)"`, we see why linkage is broken:
```
total 12
drwxr-xr-x   3 runner docker 4096 Sep  9 17:37 .
drwxr-xr-x+ 13 runner docker 4096 Sep  9 17:37 ..
drwxr-xr-x   8 runner docker 4096 Sep  9 17:37 11.3.0.reinstall
```
I get the same results when running the same tests inside a Docker container. ([workflow](https://github.com/carlocab/workflow-test/blob/585fafb6d60139e69d7b5a62ba42a784ed62f7a1/.github/workflows/test.yml))

**Comment by Bo98** [comment]
Thanks, I can reproduce the issue and am looking into it.

> 2\. The source rebuild should probably either not abort after moving the `prefix` to `"#{prefix}.reinstall"`, or put it back after aborting.

This should be a separate bug report. I'm looking into libstdc++ & libgcc_s linkage issue specifically.

**Comment by Bo98** [comment]
Fix released in 3.6.1. Let me know if you still have any issues when you build your image.
